File: src/utils/event-diagnosis.ts
import type { TenantBrand } from '../config/schema';

/**
 * 事件订阅自动诊断（research/01 的「最大增量点」）。
 *
 * 飞书对「添加事件 / 回调 / 发布版本」没有任何**写入类** OpenAPI，但提供了只读的
 * 「获取应用版本列表」（`GET /application/v6/applications/:app_id/app_versions`），
 * 响应里每个版本都带 `events`（已订阅事件列表）与 `status`（1=审核通过/已上架）。
 * 凭它可把「@机器人没反应」从"等用户发现"变成启动/doctor/诊断卡上的精确三态：
 *
 *   - `unpublished` —— 从未发布过版本（事件订阅尚未生效）；
 *   - `missing`     —— 已发布但缺 {@link REQUIRED_EVENTS}（im.message.receive_v1）；
 *   - `ok`          —— 必需事件齐全（可选事件单列，不影响状态）。
 *
 * 第四态 `unchecked` 是优雅降级：缺 `application:application.app_version:readonly`
 * scope、网络不通、接口报错都归这里——只告知、绝不阻塞启动（项目既有策略）。
 *
 * 注意：`events` 只含「事件配置」标签页的事件；「回调配置」（card.action.trigger
 * 卡片回传交互）不在其中，**无法检测**——相关提示仍须保留人工指引。
 */

const ENDPOINTS: Record<TenantBrand, string> = {
  feishu: 'https://open.feishu.cn',
  lark: 'https://open.larksuite.com',
};

/** 核心链路必需的事件：缺了 = @bot / 私聊都收不到消息。 */
export const REQUIRED_EVENTS = ['im.message.receive_v1'] as const;

/**
 * 可选功能各自依赖的事件：缺了只是对应功能静默关闭（机器人菜单 / 文档评论回复 /
 * 加入存量群绑定与自动解绑 / 表情回复驱动），不影响诊断状态，仅在结果里单列提醒。
 */
export const OPTIONAL_EVENTS = [
  'application.bot.menu_v6',
  'drive.notice.comment_add_v1',
  'im.chat.member.bot.added_v1',
  'im.chat.member.bot.deleted_v1',
  'im.message.reaction.created_v1',
] as const;

export type EventDiagnosisState = 'unchecked' | 'unpublished' | 'missing' | 'ok';

export interface EventDiagnosis {
  state: EventDiagnosisState;
  /** unchecked：没查成的原因（缺 scope / 网络 / 非 0 错误码）。 */
  reason?: string;
  /** 最新已上架（status=1）版本的版本号，如 "1.0.0"。 */
  version?: string;
  /** 该版本已订阅的全部事件。 */
  events?: string[];
  /** {@link REQUIRED_EVENTS} 中缺失的（`state === 'missing'` 时非空）。 */
  missingRequired?: string[];
  /** {@link OPTIONAL_EVENTS} 中缺失的（不影响 state）。 */
  missingOptional?: string[];
}

interface TokenResp {
  code?: number;
  msg?: string;
  tenant_access_token?: string;
}

interface VersionListResp {
  code?: number;
  msg?: string;
  data?: { items?: { version?: string; status?: number; events?: string[] }[] };
}

/**
 * 拉取应用版本列表并诊断事件订阅状态。任何失败（含缺 scope 的非 0 错误码）都
 * 落到 `unchecked` 并带原因——绝不 throw，调用方可无脑织进日志/卡片。
 * `fetchFn` 仅供测试注入。
 */
export async function diagnoseEventSubscription(
  appId: string,
  appSecret: string,
  tenant: TenantBrand,
  fetchFn: typeof fetch = fetch,
): Promise<EventDiagnosis> {
  const base = ENDPOINTS[tenant];
  let token: string;
  try {
    const resp = await fetchFn(`${base}/open-apis/auth/v3/tenant_access_token/internal`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ app_id: appId, app_secret: appSecret }),
    });
    if (!resp.ok) return { state: 'unchecked', reason: `token HTTP ${resp.status}` };
    const data = (await resp.json()) as TokenResp;
    if (data.code !== 0 || !data.tenant_access_token) {
      return { state: 'unchecked', reason: `token code=${data.code ?? '?'} msg=${data.msg ?? '<no msg>'}` };
    }
    token = data.tenant_access_token;
  } catch (err) {
    return { state: 'unchecked', reason: `网络错误：${err instanceof Error ? err.message : String(err)}` };
  }

  // order=0 = 按时间倒序 → 第一个 status=1 的就是当前在线（最新已上架）版本。
  let body: VersionListResp;
  try {
    const resp = await fetchFn(
      `${base}/open-apis/application/v6/applications/${encodeURIComponent(appId)}/app_versions?lang=zh_cn&page_size=50&order=0`,
      { headers: { Authorization: `Bearer ${token}` } },
    );
    // 飞书即便 HTTP 4xx 也常带 {code,msg}（如缺 scope 返回 400 + 99991672）——先尝试读 body
    // 拿到可读原因，而不是甩一个裸状态码。读不到 body 才回退到 HTTP 状态。
    body = (await resp.json().catch(() => undefined)) as VersionListResp | undefined ?? { code: -1, msg: '' };
    if (!resp.ok && body.code === -1) {
      const hint = resp.status === 400 || resp.status === 403 ? '——可能缺 application:application.app_version:readonly 权限' : '';
      return { state: 'unchecked', reason: `HTTP ${resp.status}${hint}` };
    }
  } catch (err) {
    return { state: 'unchecked', reason: `网络错误：${err instanceof Error ? err.message : String(err)}` };
  }
  if (body.code !== 0) {
    // 典型：缺 application:application.app_version:readonly scope → 非 0 错误码。
    const scopeHint = body.code === 99991672 || /permission|scope|access/i.test(body.msg ?? '')
      ? '——请在「权限管理」授权 application:application.app_version:readonly 后重试'
      : '';
    return { state: 'unchecked', reason: `code=${body.code ?? '?'} msg=${body.msg ?? '<no msg>'}${scopeHint}` };
  }

  const live = (body.data?.items ?? []).find((v) => v.status === 1);
  if (!live) return { state: 'unpublished' };

  const events = live.events ?? [];
  const has = new Set(events);
  const missingRequired = REQUIRED_EVENTS.filter((e) => !has.has(e));
  const missingOptional = OPTIONAL_EVENTS.filter((e) => !has.has(e));
  return {
    state: missingRequired.length > 0 ? 'missing' : 'ok',
    version: live.version,
    events,
    missingRequired,
    missingOptional,
  };
}

/** 一行中文摘要，供启动日志 / doctor CLI 直接打印（卡片自己渲染富文本）。 */
export function summarizeEventDiagnosis(d: EventDiagnosis): string {
  switch (d.state) {
    case 'ok':
      return `✅ 已生效（版本 v${d.version ?? '?'} 已订阅 ${REQUIRED_EVENTS.join(' / ')}）`;
    case 'missing':
      return `❌ 已发布版本 v${d.version ?? '?'} 缺事件：${(d.missingRequired ?? []).join('、')} —— @我 不会有反应`;
    case 'unpublished':
      return '❌ 从未发布过版本 —— 事件订阅尚未生效，@我 不会有反应';
    case 'unchecked':
      return `⚠️ 未能自动检测（${d.reason ?? '未知原因'}）`;
  }
}

/**
 * 轮询版本 API 直到事件订阅生效（state='ok'）或超时。run 用它在用户照深链配完
 * 「事件配置 + 发布版本」后主动播报「事件已生效」（research/01 建议 6 的闭环确认）。
 * 返回 ok 的诊断结果；超时返回 null。绝不 throw。
 */
export async function pollEventSubscription(
  appId: string,
  appSecret: string,
  tenant: TenantBrand,
  opts: { intervalMs?: number; timeoutMs?: number; fetchFn?: typeof fetch } = {},
): Promise<EventDiagnosis | null> {
  const intervalMs = opts.intervalMs ?? 15_000;
  const timeoutMs = opts.timeoutMs ?? 10 * 60_000;
  const deadline = Date.now() + timeoutMs;
  for (;;) {
    const d = await diagnoseEventSubscription(appId, appSecret, tenant, opts.fetchFn ?? fetch);
    if (d.state === 'ok') return d;
    if (Date.now() + intervalMs > deadline) return null;
    await new Promise((r) => setTimeout(r, intervalMs));
  }
}

File: internal/utils/eventdiagnosis.go
package utils

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"

	"github.com/modelzen/feishu-codex-bridge/internal/config"
)

// eventdiagnosis.go —— 事件订阅三态诊断（对齐 TS utils/event-diagnosis）。
//
// 飞书对「添加事件/回调/发布版本」没有写入 API，但提供只读「应用版本列表」：
// 每个版本带 events（已订阅事件）与 status（1=已上架）。据此把「@bot 没反应」从
// 「等用户发现」变成精确三态 + unchecked 优雅降级（绝不阻塞启动）。

// REQUIRED_EVENTS 核心链路必需事件。
var REQUIRED_EVENTS = []string{"im.message.receive_v1"}

// OPTIONAL_EVENTS 可选功能依赖的事件（缺失只关对应功能，不影响诊断状态）。
var OPTIONAL_EVENTS = []string{
	"application.bot.menu_v6",
	"drive.notice.comment_add_v1",
	"im.chat.member.bot.added_v1",
	"im.chat.member.bot.deleted_v1",
	"im.message.reaction.created_v1",
}

// EventDiagnosisState 四态。
type EventDiagnosisState string

const (
	EventDiagnosisUnchecked   EventDiagnosisState = "unchecked"
	EventDiagnosisUnpublished EventDiagnosisState = "unpublished"
	EventDiagnosisMissing     EventDiagnosisState = "missing"
	EventDiagnosisOK          EventDiagnosisState = "ok"
)

// EventDiagnosis 诊断结果。
type EventDiagnosis struct {
	State           EventDiagnosisState
	Reason          string
	Version         string
	Events          []string
	MissingRequired []string
	MissingOptional []string
}

// DiagnoseEventSubscription 用默认 base 诊断。
func DiagnoseEventSubscription(ctx context.Context, appID, appSecret string, tenant config.TenantBrand, hc *http.Client) EventDiagnosis {
	return diagnoseAt(ctx, hc, tenantBase(tenant), appID, appSecret)
}

func diagnoseAt(ctx context.Context, hc *http.Client, base, appID, appSecret string) EventDiagnosis {
	if hc == nil {
		hc = http.DefaultClient
	}
	// 1. token
	token, reason := fetchTenantToken(ctx, hc, base, appID, appSecret)
	if reason != "" {
		return EventDiagnosis{State: EventDiagnosisUnchecked, Reason: reason}
	}
	// 2. 版本列表
	code, body, err := getJSON(ctx, hc, fmt.Sprintf("%s/open-apis/application/v6/applications/%s/app_versions?lang=zh_cn&page_size=50&order=1",
		base, url.PathEscape(appID)), "Bearer "+token)
	if err != nil {
		return EventDiagnosis{State: EventDiagnosisUnchecked, Reason: "网络错误：" + err.Error()}
	}
	var vr versionListResp
	parseErr := json.Unmarshal(body, &vr)
	if code != http.StatusOK {
		hint := ""
		if (code == http.StatusBadRequest || code == http.StatusForbidden) && parseErr != nil {
			hint = "——可能缺 application:application.app_version:readonly 权限"
		}
		if parseErr != nil {
			return EventDiagnosis{State: EventDiagnosisUnchecked, Reason: fmt.Sprintf("HTTP %d%s", code, hint)}
		}
	}
	if parseErr != nil {
		return EventDiagnosis{State: EventDiagnosisUnchecked, Reason: "响应不是合法 JSON"}
	}
	if vr.Code != 0 {
		scopeHint := ""
		if vr.Code == 99991672 || scopePermissionRegexp.MatchString(vr.Msg) {
			scopeHint = "——请在「权限管理」授权 application:application.app_version:readonly 后重试"
		}
		return EventDiagnosis{State: EventDiagnosisUnchecked, Reason: fmt.Sprintf("code=%d msg=%s%s", vr.Code, fallback(vr.Msg, "<no msg>"), scopeHint)}
	}
	// 收集所有已上架（status=1）版本，按语义版本号取「最新」作为当前在线版本。
	// 注意：飞书 app_versions 的 order 参数语义在不同端并不一致（文档写 0=升序，
	// 旧注释误以为 0=倒序），且可能同时返回多个 status=1。故这里不依赖接口排序，
	// 自行用语义版本比较选出最新已上架版本——否则会误判「旧版本缺事件」而漏掉
	// 真正在线、已含事件的新版本（表现为「明明已配事件却报缺」）。
	var live *versionItem
	for i := range vr.Data.Items {
		if vr.Data.Items[i].Status != 1 {
			continue
		}
		if live == nil || semverGreater(vr.Data.Items[i].Version, live.Version) {
			live = &vr.Data.Items[i]
		}
	}
	if live == nil {
		return EventDiagnosis{State: EventDiagnosisUnpublished}
	}
	events := live.Events
	has := map[string]bool{}
	for _, e := range events {
		has[e] = true
	}
	// 飞书 app_versions 的 events 字段是「中文展示名」（如 "接收消息"），
	// 与 REQUIRED/OPTIONAL_EVENTS 的事件码（如 im.message.receive_v1）不匹配，
	// 直接比较会误判「明明已配事件却报缺」。故额外收集 event_infos[].event_type
	//（真实事件码）用于判定。
	for _, ei := range live.EventInfos {
		if ei.EventType != "" {
			has[ei.EventType] = true
		}
	}
	missingReq := diffEvents(has, REQUIRED_EVENTS)
	missingOpt := diffEvents(has, OPTIONAL_EVENTS)
	state := EventDiagnosisOK
	if len(missingReq) > 0 {
		state = EventDiagnosisMissing
	}
	return EventDiagnosis{
		State:           state,
		Version:         live.Version,
		Events:          events,
		MissingRequired: missingReq,
		MissingOptional: missingOpt,
	}
}

var scopePermissionRegexp = stringsNewRegexp(`(?i)permission|scope|access`)

// semverGreater 判断版本号 a 是否比 b 更大（如 1.0.10 > 1.0.3）。
// 按 "." 拆段逐段数值比较；非数字段退化为字符串比较；忽略前缀 "v"。
func semverGreater(a, b string) bool {
	pa := splitVersion(a)
	pb := splitVersion(b)
	n := len(pa)
	if len(pb) > n {
		n = len(pb)
	}
	for i := 0; i < n; i++ {
		sa, sb := "", ""
		if i < len(pa) {
			sa = pa[i]
		}
		if i < len(pb) {
			sb = pb[i]
		}
		na, erra := strconv.Atoi(sa)
		nb, errb := strconv.Atoi(sb)
		if erra == nil && errb == nil {
			if na != nb {
				return na > nb
			}
			continue
		}
		if sa != sb {
			return sa > sb
		}
	}
	return false
}

func splitVersion(v string) []string {
	return strings.Split(strings.TrimPrefix(v, "v"), ".")
}

func diffEvents(has map[string]bool, list []string) []string {
	out := []string{}
	for _, e := range list {
		if !has[e] {
			out = append(out, e)
		}
	}
	return out
}

type versionListResp struct {
	Code int    `json:"code"`
	Msg  string `json:"msg"`
	Data struct {
		Items []versionItem `json:"items"`
	} `json:"data"`
}

type versionItem struct {
	Version    string      `json:"version"`
	Status     int         `json:"status"`
	Events     []string    `json:"events"`      // 中文展示名（如 "接收消息"）
	EventInfos []eventInfo `json:"event_infos"` // 真实事件码（如 im.message.receive_v1）
}

type eventInfo struct {
	EventName string `json:"event_name"`
	EventType string `json:"event_type"`
}

// SummarizeEventDiagnosis 一行中文摘要（启动日志 / doctor CLI 用）。
func SummarizeEventDiagnosis(d EventDiagnosis) string {
	switch d.State {
	case EventDiagnosisOK:
		return fmt.Sprintf("✅ 已生效（版本 v%s 已订阅 %s）", fallback(d.Version, "?"), strings.Join(REQUIRED_EVENTS, " / "))
	case EventDiagnosisMissing:
		return fmt.Sprintf("❌ 已发布版本 v%s 缺事件：%s —— @我 不会有反应", fallback(d.Version, "?"), strings.Join(d.MissingRequired, "、"))
	case EventDiagnosisUnpublished:
		return "❌ 从未发布过版本 —— 事件订阅尚未生效，@我 不会有反应"
	default:
		return fmt.Sprintf("⚠️ 未能自动检测（%s）", fallback(d.Reason, "未知原因"))
	}
}

// PollEventSubscription 轮询直到 state=ok 或超时；超时返回 nil。
func PollEventSubscription(ctx context.Context, appID, appSecret string, tenant config.TenantBrand, hc *http.Client, interval, timeout time.Duration) *EventDiagnosis {
	if interval <= 0 {
		interval = 15 * time.Second
	}
	if timeout <= 0 {
		timeout = 10 * time.Minute
	}
	deadline := time.Now().Add(timeout)
	for {
		d := DiagnoseEventSubscription(ctx, appID, appSecret, tenant, hc)
		if d.State == EventDiagnosisOK {
			return &d
		}
		if time.Now().Add(interval).After(deadline) {
			return nil
		}
		select {
		case <-ctx.Done():
			return nil
		case <-time.After(interval):
		}
	}
}
